File: capacitor-plugins/capacitor-spotify/ios/Plugin/SpotifyPlugin.swift
import Foundation
import Capacitor
import SpotifyiOS

@objc(SpotifyPlugin)
public class SpotifyPlugin: CAPPlugin, CAPBridgedPlugin, SPTAppRemoteDelegate, SPTAppRemotePlayerStateDelegate {
    public let identifier = "SpotifyPlugin"
    public let jsName = "Spotify"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isConnected", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "play", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resume", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "next", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "previous", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "seek", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPlayerState", returnType: CAPPluginReturnPromise),
    ]

    private var appRemote: SPTAppRemote?
    private var connectCall: CAPPluginCall?
    private var clientId: String = ""
    private var redirectUri: String = ""

    override public func load() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleSpotifyCallback(_:)),
            name: NSNotification.Name("SpotifyCallback"),
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleAppDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
    }

    @objc func initialize(_ call: CAPPluginCall) {
        guard let clientId = call.getString("clientId"),
              let redirectUri = call.getString("redirectUri") else {
            call.reject("clientId and redirectUri are required")
            return
        }
        self.clientId = clientId
        self.redirectUri = redirectUri

        let configuration = SPTConfiguration(clientID: clientId, redirectURL: URL(string: redirectUri)!)
        appRemote = SPTAppRemote(configuration: configuration, logLevel: .debug)
        appRemote?.delegate = self

        call.resolve()
    }

    @objc func connect(_ call: CAPPluginCall) {
        guard let appRemote = appRemote else {
            call.reject("Plugin not initialized. Call initialize() first.")
            return
        }

        connectCall = call

        // Set access token placeholder to trigger auth
        appRemote.connectionParameters.accessToken = ""
        appRemote.authorizeAndPlayURI("")

        // The actual connection happens in handleSpotifyCallback after OAuth redirect
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        appRemote?.disconnect()
        call.resolve()
    }

    @objc func isConnected(_ call: CAPPluginCall) {
        call.resolve(["connected": appRemote?.isConnected ?? false])
    }

    @objc func play(_ call: CAPPluginCall) {
        guard let uri = call.getString("uri") else {
            call.reject("uri is required")
            return
        }
        guard let playerAPI = appRemote?.playerAPI else {
            call.reject("Not connected to Spotify")
            return
        }

        playerAPI.play(uri) { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                if let positionMs = call.getInt("positionMs"), positionMs > 0 {
                    playerAPI.seek(toPosition: positionMs) { _, _ in }
                }
                call.resolve()
            }
        }
    }

    @objc func resume(_ call: CAPPluginCall) {
        appRemote?.playerAPI?.resume { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func pause(_ call: CAPPluginCall) {
        appRemote?.playerAPI?.pause { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func next(_ call: CAPPluginCall) {
        appRemote?.playerAPI?.skip(toNext: { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        })
    }

    @objc func previous(_ call: CAPPluginCall) {
        appRemote?.playerAPI?.skip(toPrevious: { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        })
    }

    @objc func seek(_ call: CAPPluginCall) {
        guard let positionMs = call.getInt("positionMs") else {
            call.reject("positionMs is required")
            return
        }
        appRemote?.playerAPI?.seek(toPosition: positionMs) { _, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve()
            }
        }
    }

    @objc func getPlayerState(_ call: CAPPluginCall) {
        appRemote?.playerAPI?.getPlayerState { state, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            guard let state = state as? SPTAppRemotePlayerState else {
                call.reject("Could not get player state")
                return
            }
            call.resolve(self.playerStateToDict(state))
        }
    }

    // MARK: - SPTAppRemoteDelegate

    public func appRemoteDidEstablishConnection(_ appRemote: SPTAppRemote) {
        appRemote.playerAPI?.delegate = self
        appRemote.playerAPI?.subscribe(toPlayerState: { _, _ in })

        // Resolve pending connect call
        if let call = connectCall {
            call.resolve(["displayName": "Spotify User"])
            connectCall = nil
        }

        notifyListeners("connectionStatusChanged", data: ["connected": true])
    }

    public func appRemote(_ appRemote: SPTAppRemote, didFailConnectionAttemptWithError error: Error?) {
        if let call = connectCall {
            call.reject(error?.localizedDescription ?? "Connection failed")
            connectCall = nil
        }
        notifyListeners("connectionStatusChanged", data: ["connected": false])
    }

    public func appRemote(_ appRemote: SPTAppRemote, didDisconnectWithError error: Error?) {
        notifyListeners("connectionStatusChanged", data: ["connected": false])
    }

    // MARK: - SPTAppRemotePlayerStateDelegate

    public func playerStateDidChange(_ playerState: SPTAppRemotePlayerState) {
        notifyListeners("playerStateChanged", data: playerStateToDict(playerState))
    }

    // MARK: - Helpers

    @objc private func handleSpotifyCallback(_ notification: Notification) {
        guard let url = notification.object as? URL else { return }

        let params = appRemote?.authorizationParameters(from: url)
        if let accessToken = params?[SPTAppRemoteAccessTokenKey] {
            appRemote?.connectionParameters.accessToken = accessToken
            appRemote?.connect()
        } else if let error = params?[SPTAppRemoteErrorDescriptionKey] {
            if let call = connectCall {
                call.reject(error)
                connectCall = nil
            }
        }
    }

    @objc private func handleAppDidBecomeActive() {
        if let appRemote = appRemote,
           !appRemote.isConnected,
           appRemote.connectionParameters.accessToken != nil,
           !appRemote.connectionParameters.accessToken!.isEmpty {
            appRemote.connect()
        }
    }

    private func playerStateToDict(_ state: SPTAppRemotePlayerState) -> [String: Any] {
        return [
            "trackName": state.track.name,
            "artistName": state.track.artist.name,
            "albumName": state.track.album.name,
            "albumArtUrl": "",
            "durationMs": state.track.duration,
            "positionMs": state.playbackPosition,
            "isPaused": state.isPaused,
            "uri": state.track.uri,
        ]
    }
}

File: capacitor-plugins/capacitor-spotify/android/src/main/java/com/cardio/spotify/SpotifyPlugin.java
package com.cardio.spotify;

import android.app.Activity;
import android.content.Intent;

import androidx.activity.result.ActivityResult;
import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.spotify.android.appremote.api.ConnectionParams;
import com.spotify.android.appremote.api.Connector;
import com.spotify.android.appremote.api.SpotifyAppRemote;
import com.spotify.protocol.types.PlayerState;
import com.spotify.protocol.types.Track;
import com.spotify.sdk.android.auth.AuthorizationClient;
import com.spotify.sdk.android.auth.AuthorizationRequest;
import com.spotify.sdk.android.auth.AuthorizationResponse;

@CapacitorPlugin(name = "Spotify")
public class SpotifyPlugin extends Plugin {

    private SpotifyAppRemote appRemote;
    private String clientId = "";
    private String redirectUri = "";
    private PluginCall pendingConnectCall;

    @PluginMethod
    public void initialize(PluginCall call) {
        clientId = call.getString("clientId", "");
        redirectUri = call.getString("redirectUri", "");

        if (clientId.isEmpty() || redirectUri.isEmpty()) {
            call.reject("clientId and redirectUri are required");
            return;
        }
        call.resolve();
    }

    @PluginMethod
    public void connect(PluginCall call) {
        if (clientId.isEmpty()) {
            call.reject("Plugin not initialized. Call initialize() first.");
            return;
        }

        pendingConnectCall = call;

        AuthorizationRequest.Builder builder = new AuthorizationRequest.Builder(
                clientId,
                AuthorizationResponse.Type.TOKEN,
                redirectUri
        );
        builder.setScopes(new String[]{
                "app-remote-control",
                "user-modify-playback-state",
                "user-read-playback-state",
                "user-read-currently-playing"
        });

        Intent authIntent = AuthorizationClient.createLoginActivityIntent(
                getActivity(),
                builder.build()
        );

        startActivityForResult(call, authIntent, "handleAuthResult");
    }

    @ActivityCallback
    private void handleAuthResult(PluginCall call, ActivityResult result) {
        if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
            AuthorizationResponse response = AuthorizationClient.getResponse(
                    result.getResultCode(),
                    result.getData()
            );

            if (response.getType() == AuthorizationResponse.Type.TOKEN) {
                connectToAppRemote(response.getAccessToken(), call);
            } else if (response.getType() == AuthorizationResponse.Type.ERROR) {
                call.reject("Authorization error: " + response.getError());
            } else {
                call.reject("Authorization cancelled");
            }
        } else {
            call.reject("Authorization cancelled");
        }
    }

    private void connectToAppRemote(String accessToken, PluginCall call) {
        ConnectionParams params = new ConnectionParams.Builder(clientId)
                .setRedirectUri(redirectUri)
                .showAuthView(false)
                .build();

        SpotifyAppRemote.connect(getContext(), params, new Connector.ConnectionListener() {
            @Override
            public void onConnected(SpotifyAppRemote remote) {
                appRemote = remote;
                subscribeToPlayerState();

                JSObject result = new JSObject();
                result.put("displayName", "Spotify User");
                call.resolve(result);

                JSObject status = new JSObject();
                status.put("connected", true);
                notifyListeners("connectionStatusChanged", status);
            }

            @Override
            public void onFailure(Throwable error) {
                call.reject("Connection failed: " + error.getMessage());

                JSObject status = new JSObject();
                status.put("connected", false);
                notifyListeners("connectionStatusChanged", status);
            }
        });
    }

    private void subscribeToPlayerState() {
        if (appRemote == null) return;

        appRemote.getPlayerApi().subscribeToPlayerState().setEventCallback(playerState -> {
            notifyListeners("playerStateChanged", playerStateToJS(playerState));
        });
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        if (appRemote != null) {
            SpotifyAppRemote.disconnect(appRemote);
            appRemote = null;
        }

        JSObject status = new JSObject();
        status.put("connected", false);
        notifyListeners("connectionStatusChanged", status);

        call.resolve();
    }

    @PluginMethod
    public void isConnected(PluginCall call) {
        JSObject result = new JSObject();
        result.put("connected", appRemote != null && appRemote.isConnected());
        call.resolve(result);
    }

    @PluginMethod
    public void play(PluginCall call) {
        if (!ensureConnected(call)) return;

        String uri = call.getString("uri");
        if (uri == null || uri.isEmpty()) {
            call.reject("uri is required");
            return;
        }

        appRemote.getPlayerApi().play(uri).setResultCallback(empty -> {
            Integer positionMs = call.getInt("positionMs");
            if (positionMs != null && positionMs > 0) {
                appRemote.getPlayerApi().seekTo(positionMs);
            }
            call.resolve();
        }).setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void resume(PluginCall call) {
        if (!ensureConnected(call)) return;
        appRemote.getPlayerApi().resume()
                .setResultCallback(empty -> call.resolve())
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void pause(PluginCall call) {
        if (!ensureConnected(call)) return;
        appRemote.getPlayerApi().pause()
                .setResultCallback(empty -> call.resolve())
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void next(PluginCall call) {
        if (!ensureConnected(call)) return;
        appRemote.getPlayerApi().skipNext()
                .setResultCallback(empty -> call.resolve())
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void previous(PluginCall call) {
        if (!ensureConnected(call)) return;
        appRemote.getPlayerApi().skipPrevious()
                .setResultCallback(empty -> call.resolve())
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void seek(PluginCall call) {
        if (!ensureConnected(call)) return;
        Integer positionMs = call.getInt("positionMs");
        if (positionMs == null) {
            call.reject("positionMs is required");
            return;
        }
        appRemote.getPlayerApi().seekTo(positionMs)
                .setResultCallback(empty -> call.resolve())
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    @PluginMethod
    public void getPlayerState(PluginCall call) {
        if (!ensureConnected(call)) return;
        appRemote.getPlayerApi().getPlayerState()
                .setResultCallback(state -> call.resolve(playerStateToJS(state)))
                .setErrorCallback(error -> call.reject(error.getMessage()));
    }

    private boolean ensureConnected(PluginCall call) {
        if (appRemote == null || !appRemote.isConnected()) {
            call.reject("Not connected to Spotify");
            return false;
        }
        return true;
    }

    private JSObject playerStateToJS(@NonNull PlayerState state) {
        JSObject obj = new JSObject();
        Track track = state.track;
        obj.put("trackName", track != null ? track.name : "");
        obj.put("artistName", track != null && track.artist != null ? track.artist.name : "");
        obj.put("albumName", track != null && track.album != null ? track.album.name : "");
        obj.put("albumArtUrl", "");
        obj.put("durationMs", track != null ? track.duration : 0);
        obj.put("positionMs", state.playbackPosition);
        obj.put("isPaused", state.isPaused);
        obj.put("uri", track != null ? track.uri : "");
        return obj;
    }
}
